### ISSUE TITLE
vkCreateSwapchainKHR imageExtent surface capabilities false positive

### ISSUE THREAD

**Issue by Ralith** [KhronosGroup/Vulkan-ValidationLayers#1340]
`CoreChecks::ValidateCreateSwapchain` checks that the supplied `imageExtent` falls within the bounds given by `vkGetPhysicalDeviceSurfaceCapabilitiesKHR`. However, it's possible that a correct application will call `vkGetPhysicalDeviceSurfaceCapabilitiesKHR`, followed by the underlying surface being resized before the call to `vkCreateSwapchainKHR`, causing the layer's own call to produce different results. To ensure consistency, the layer should probably cache the results from the most recent call made by the application, if any.

**Comment by krOoze** [comment]
Actually that seems correct-ish to me. The app should prevent the window to change size by platform API means before creating the swapchain. If it fails to do so, it would still be invalid use.

E.g. consider calling `vkGetPhysicalDeviceSurfaceCapabilitiesKHR` only once, then handling several `WM_SIZE`. The original value is obviously long invalid. Probably needs to be clarified in spec. This probably is relevant (4th paragraph): https://github.com/KhronosGroup/Vulkan-Docs/issues/388

**Comment by Ralith** [comment]
> The app should prevent the window to change size by platform API means before creating the swapchain.

This is impossible in some environments, e.g. X11 with a tiling window manager.

**Comment by krOoze** [comment]
That may be so, but if the platform can change size willy-nilly, then it should probably not be a VU, but a runtime error.

**Comment by Ralith** [comment]
The spec says:

> imageExtent is the size (in pixels) of the swapchain image(s). The behavior is platform-dependent if the image extent does not match the surface’s currentExtent as returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR.

Note "platform-dependent" rather than "undefined." I think the most reasonable behavior here is to return SUBOPTIMAL or OUT_OF_DATE from subsequent acquire/present operations so an application's normal resize path can take effect. I don't see a good way for swapchain creation to fail outright in a way that the application can easily determine is recoverable.

**Comment by krOoze** [comment]
The quote means something else. It says it is platform dependent whether the image stretches or whatever else. It does not say if creating such swapchain is allowed in the first place. And it is not for Xlib (, XCB, Win32).

There's also `VK_ERROR_INITIALIZATION_FAILED`, though guaranteed `OUT_OF_DATE` would be nice. (That is if we accept in the first place that `vkCreateSwapchain` can accept even old size values.)

**Comment by Ralith** [comment]
It says "the behavior is platform-dependent," period, and creating such a swapchain seems to work more or less just as I described here.

Returning `VK_ERROR_INITIALIZATION_FAILED` is not useful behavior, because the application cannot reasonably judge that it should retry.

**Comment by haasn** [comment]
Note: Rather than removing the check completely, it would suffice to merely relax the parameter check to only enforce `minImageExtent` and `maxImageExtent` if they differ from `currentExtent`. That way, the parameter validation still works as intended for platforms such as wayland (on which they represent true hardware limits), but effectively disables this parameter validation for platforms such as x11/win32 on which those three fields are specified to be identical.

**Comment by krOoze** [comment]
@haasn That assumes `maxImageExtent` cannot itself change :p.

> (And possibly there might also need to be a special case for the situation in which min = max = current = (0, 0), in which case any value of imageExtent is illegal)

It is a separate VU, so I don't see a problem there...
Stats say it is implemented.

**Comment by mark-lunarg** [comment]
@krOoze, @Ralith, just to be clear, is the VU to removed `VUID-VkSwapchainCreateInfoKHR-imageExtent-01274`?

**Comment by krOoze** [comment]
Well, yea. Assuming that is da wei. Follow https://github.com/KhronosGroup/Vulkan-Docs/issues/1144

**Comment by mark-lunarg** [comment]
Assuming @jekstrand's proposed interpretation is correct, and even though the layers seem to match the spec behavior, planning to remove this check.

**Comment by mark-lunarg** [comment]
Is it worth moving this to Best Practices?

**Comment by jekstrand** [comment]
That's an interesting question.  Honestly, I think it's probably a window-system-specific question in some ways.  On Wayland, the minimum is 1x1 and the maximum is likely the maximum image size for a 2D image.  In that case, it's never going to change, there is no race, and it's a valid VU to be validating.

On other platforms, the result is far more dynamic and may have to do with the size of the screen or the size of the window.  In that case, I think the VU still stands because it's saying "query this thing and then pass a value between these two numbers" which is something that the application must do to the best of its ability.  The problem is that it's racy and therefore virtually impossible to validate.  The best thing you could do would be to update some field on the surface every time the client does the query and then check to see if it matches the last one rather than the validation layers re-querying.  I'm not sure if that's practical, though.  It's also not really "correct" but it should be better.  Another option would be to catch every client query and record the maximum value ever returned by a query and simply check against that.  I don't really know.  It'd all be very heuristic and, honestly, probably not worth the time.

**Comment by krOoze** [comment]
If it is actually well defined (via `VK_ERROR_OUT_OF_DATE_KHR`), then it does not need much validation.

`minExtent` and `maxExtent` sound like a hard limit though, and should be validated. Though depends on formal resolution to https://github.com/KhronosGroup/Vulkan-Docs/issues/388.

**Comment by mark-lunarg** [comment]
`VUID-VkSwapchainCreateInfoKHR-imageExtent-01274` is implemented.

**Comment by TheNumbat** [comment]
I've also run into this issue; a validation error is triggered when the surface extent race condition occurs during resizing on X. 

According to [Vulkan-Docs/issues/828](https://github.com/KhronosGroup/Vulkan-Docs/issues/828) and what was suggested above, it seems that in this case vkCreateSwapchainKHR should create a new swapchain that immediately gets marked as out of date, since the spec doesn't define a relevant error to return. However, the validation layer is instead throwing an error, so nothing is created and there's no way to recover short of ignoring the message and recreating in a loop until the resize ends. 

Is this a bug in the validation layer?

**Comment by kaiserschmarrn0** [comment]
> Wasn't it implemented all along? The issue here is a validation error being generated in circumstances that an application cannot possibly prevent.

with xwayland it is even more of an issue, the race condition is triggered much more often and it seems to be fatal. that said, xwayland is probably not a priority when vulkan supports wayland native.

**Comment by Ralith** [comment]
@mark-lunarg, please reopen the issue; this bug still exists.

**Comment by mark-lunarg** [comment]
Are y'all getting the 01274 error in these cases?  I looks like the only thing validation does there is to throw an error, but this should not affect creating the swapchain in any way (unless TRUE is returned from the debug callback).  The spec seems pretty clear in this case and if the check is removed it wouldn't catch errors for folks who _are_ doing it incorrectly (which we completely understand, is **not** you guys).   It looks like [vulkan-docs hasn't budged](https://github.com/KhronosGroup/Vulkan-Docs/issues/1144) here.

**Comment by TheNumbat** [comment]
Yeah, I believe swapchain creation is working as intended given I ignore the validation error. It's just a bit inelegant to have to ignore the error in this specific case, since as you said it is a real error when the usage is incorrect.

**Comment by Ralith** [comment]
I would strongly prefer that the validation layers not knowingly emit false positives. That undermine the usefulness of the validation layers as a whole by training people to ignore them, and is a recurring source of pain and confusion for new users, who must rediscover this issue the hard way.

In the original post I described a compromise solution that should still catch severe bugs, but even disabling the validation entirely until upstream addresses the issue would be an improvement.

**Comment by sfricke-samsung** [comment]
So I just saw this issue and I actually a bit ago tried to add support for multi-surface tracking of surface capabilities... so uploaded to #2573 as it can be used as a base for this issue as I currently had the issue of knowing "when" to have the layer go and query for the newest capabilities

**Comment by buzmeg** [comment]
Just hit this as well on Windows 10 and nVidia.

Windows: 10.0.19042.928
Vulkan:  1.2.141
nVidia: GeForce RTX 2060

I'm literally going from vkGetPhysicalDeviceSurfaceCapabilitiesKHR to vkCreateSwapchainKHR with absolutely no other Vulkan calls in between.

```
Extents check: VkExtent2D { width: 1258, height: 828 } VkExtent2D { width: 1258, height: 828 } VkExtent2D { width: 1258, height: 828 }
VUID-VkSwapchainCreateInfoKHR-imageExtent-01274(ERROR / SPEC): msgNum: 2094043421 - Validation Error: [ VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 ] Object 0: handle = 0x20159966448, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x7cd0911d | vkCreateSwapchainKHR() called with imageExtent = (1258,828), which is outside the bounds returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(): currentExtent = (1267,829), minImageExtent = (1267,829), maxImageExtent = (1267,829). The Vulkan spec states: imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface (https://vulkan.lunarg.com/doc/view/1.2.170.0/windows/1.2-extensions/vkspec.html#VUID-VkSwapchainCreateInfoKHR-imageExtent-01274)
    Objects: 1
        [0] 0x20159966448, type: 3, name: NULL
```

**Comment by Ralith** [comment]
This continues to bite people regularly (see above, etc). Can the broken check be removed?

**Comment by CarloWood** [comment]
Just run into this as well. I was like WHAT?! The only thing clear to me was that I can't do anything about it. But it feels very uncomfortable not knowing the effect of plainly ignoring an ERROR from the validation layers :/.

I agree that removing the check is better than leaving it as is. A long term solution would be where the swap chain is simply created but immediately outdated (this might already be the case?). I'd say: cache the last call to `vkGetPhysicalDeviceSurfaceCapabilitiesKHR` and if the size passed to `vkCreateSwapchainKHR` is the same then no error.

That will still alert people that completely forget to call `vkGetPhysicalDeviceSurfaceCapabilitiesKHR` every time: after all, when you resize a window it will have a different size every time. So either you call `vkGetPhysicalDeviceSurfaceCapabilitiesKHR` every time, or you'll get an error imho.
